README.md

```diff
@@ -59,12 +63,143 @@ $ npm i @mixmaxhq/promise-pool
+
 * 1.1.1 Move `ava` and `ava-spec` to `devDependencies`.
 
 * 1.1.0 Adds transpilation so it can be used in Node 6 (and prior) environments.
 
 * 1.0.0 Initial release.
 
+Troubleshooting
+---------------
+
+### `cannot queue function in pool`
+
+If you're getting this error, then you're calling `start` too many times
+concurrently without waiting for the previous `start` call to resolve. Please
+read on - one of two things is happening:
+
+- you're using promise-pool in a way that will defeat its purpose
+- you're using promise-pool in an uncommon manner that works with its purpose (but might benefit from being refactored)
+
+In the former case, you probably have code that looks like this:
+
+```js
+import _ from 'lodash';
+import mongoist from 'mongoist';
+
+const db = mongoist(...);
+
+async function startJobs() {
+  const pool = new PromisePool({numConcurrent: 4});
+
+  // Pull in an array of users (high memory usage, poor performance
+  // characteristics due to loading all instead of streaming with something
+  // like promise-iterate).
+  const users = await db.users.findAsCursor().toArray();
+
+  // _.each doesn't wait on the promise returned by each invocation, so it
+  // won't apply backpressure to the loop in the manner pool.start expects.
+  _.each(users, async (user) => {
+    await pool.start(async () => {
+      await queue.publish(user);
+    });
+  });
+
+  await pool.flush();
+}
+```
+
+Instead, you need to use some iteration method that preserves backpressure, like the `for`-`of` loop:
+
+```js
+async function startJobs() {
+  const pool = new PromisePool({numConcurrent: 4});
+
+  // Still severely suboptimal.
+  const users = await db.users.findAsCursor().toArray();
+
+  for (const user of users) {
+    // Now the await applies to the `startJobs` async function instead of
+    // the anonymous async function.
+    await pool.start(async () => {
+      await queue.publish(user);
+    });
+  }
+
+  await pool.flush();
+}
+```
+
+Or even better, couple this with a call to `promise-iterate` to only load users as you need them:
+
+```js
+import {asyncIterate} from 'promise-iterate';
+
+async function startJobs() {
+  const pool = new PromisePool({numConcurrent: 4});
+
+  const users = await db.users.findAsCursor();
+
+  // promise-iterate correctly applies backpressure, and helpfully iterates
+  // the cursor without pulling in all results as an array.
+  await asyncIterate(users, async (user) => {
+    // The await still applies to the `startJobs` async function due to the
+    // behavior of promise-iterate.
+    await pool.start(async () => {
+      await queue.publish(user);
+    });
+  });
+
+  await pool.flush();
+}
+```
+
+The other case is where you're using `promise-pool` in an unorthodox manner,
+where you'll need to use `maxPending`:
+
+```js
+async function initializeAllUsers() {
+  const pool = new PromisePool({
+    numConcurrent: 4,
+
+    // Important: without this, we'll fail almost immediately after starting to
```

Comment:
Curious how would we fail immediately? (With the previous promise-pool—I know that now we would fail immediately now because of `maxPending` defaulting to 1.)

Reply:
This is a new change!